agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1411 @@
+        let mut stop_all_discovery_receiver = stop_all_discovery_sender.subscribe();
+        let task3_discovery_operator = discovery_operator.clone();
+        tasks.push(tokio::spawn(async move {
+            let kube_interface: Arc<Box<dyn k8s::KubeInterface>> = Arc::new(Box::new(k8s::create_kube_interface()));
+            loop {
+                task3_discovery_operator
+                    .check_offline_status(kube_interface.clone())
+                    .await
+                    .unwrap();
+                if tokio::time::timeout(
+                    Duration::from_secs(30),
+                    stop_all_discovery_receiver.recv(),
+                )
+                .await.is_ok()
+                {
+                    trace!("start_discovery - received message to stop checking connectivity status for configuration {}", config_name);
+                    break;
+                }
+            }
+        }));
+        futures::future::try_join_all(tasks).await?;
+        finished_all_discovery_sender.send(()).unwrap();
+        Ok(())
+    }
+
+    /// Waits to be notified of new discovery handlers. If the discovery handler does discovery for this Configuration's protocol,
+    /// discovery is kicked off.
+    async fn listen_for_new_discovery_handlers(
+        discovery_operator: Arc<DiscoveryOperator>,
+        new_discovery_handler_receiver: &mut broadcast::Receiver<String>,
+        stop_all_discovery_receiver: &mut broadcast::Receiver<()>,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        let mut tasks = Vec::new();
+        loop {
+            tokio::select! {
+                _ = try_receive(stop_all_discovery_receiver) => {
+                    trace!("listen_for_new_discovery_handlers - received message to stop discovery for configuration {}", discovery_operator.get_config().metadata.name);
+                    discovery_operator.stop_all_discovery().await;
+                    break;
+                },
+                result = try_receive(new_discovery_handler_receiver) => {
+                    // Check if it is this protocol
+                    if let Ok(protocol) = result {
+                        if protocol == discovery_operator.get_config().spec.protocol.name {
+                            trace!("listen_for_new_discovery_handlers - received new registered discovery handler for configuration {}", discovery_operator.get_config().metadata.name);
+                            let new_discovery_operator = discovery_operator.clone();
+                            tasks.push(tokio::spawn(async move {
+                                do_discover(new_discovery_operator, Arc::new(Box::new(k8s::create_kube_interface()))).await.unwrap();
+                            }));
+                        }
+                    }
+                }
+            }
+        }
+        futures::future::try_join_all(tasks).await?;
```

Comment:
for some reason, this try_join_all line feels out of place.

it looks like this code is meant to loop forever (building a list of tasks) until the stop_all signal is received ... when the stop_all signal is received, then call join_all on the tasks.

does that just make sure that this function hangs around until each of the tasks handles the stop_all_discovery message?

Reply:
> does that just make sure that this function hangs around until each of the tasks handles the stop_all_discovery message?

Yes, it is essentially to make sure that [`finished_all_discovery_sender`](https://github.com/kate-goldenring/akri/blob/0c51141cde9f03625c598c06c74079e00aeb0600/agent/src/util/discovery_operator.rs#L582) does not say it has finished until all threads have finished . A configuration isnt removed from the `ConfigMap` until it (`handle_config_delete`) receives this finished message to make sure no instances are created after its removed.

I restructured the code in 3c63d41 to hopefully make this clearer.